core/headerchain.go

```diff
@@ -0,0 +1,441 @@
+// it if present.
+func (bc *HeaderChain) HasHeader(hash common.Hash) bool {
+	return bc.GetHeader(hash) != nil
+}
+
+// GetHeaderByNumber retrieves a block header from the database by number,
+// caching it (associated with its hash) if found.
+func (self *HeaderChain) GetHeaderByNumber(number uint64) *types.Header {
+	hash := GetCanonicalHash(self.chainDb, number)
+	if hash == (common.Hash{}) {
+		return nil
+	}
+	return self.GetHeader(hash)
+}
+
+// CurrentHeader retrieves the current head header of the canonical chain. The
+// header is retrieved from the HeaderChain's internal cache.
+func (self *HeaderChain) CurrentHeader() *types.Header {
+	return self.currentHeader
+}
+
+// SetCurrentHeader sets the current head header of the canonical chain.
+func (self *HeaderChain) SetCurrentHeader(head *types.Header) {
+	if err := WriteHeadHeaderHash(self.chainDb, head.Hash()); err != nil {
+		glog.Fatalf("failed to insert head header hash: %v", err)
+	}
+	self.currentHeader = head
+}
+
+// SetHead rewinds the local chain to a new head. Everything above the new head
+// will be deleted and the new one set.
+func (bc *HeaderChain) SetHead(head uint64) {
+	height := uint64(0)
+	if bc.currentHeader != nil {
+		height = bc.currentHeader.Number.Uint64()
+	}
+
+	for bc.currentHeader != nil && bc.currentHeader.Number.Uint64() > head {
+		hash := bc.currentHeader.Hash()
+		DeleteHeader(bc.chainDb, hash)
+		DeleteTd(bc.chainDb, hash)
+		bc.currentHeader = bc.GetHeader(bc.currentHeader.ParentHash)
+	}
+	// Roll back the canonical chain numbering
+	for i := height; i > head; i-- {
+		DeleteCanonicalHash(bc.chainDb, i)
+	}
+	// Clear out any stale content from the caches
+	bc.headerCache.Purge()
+	bc.tdCache.Purge()
+
+	if bc.currentHeader == nil {
+		bc.currentHeader = bc.genesisHeader
+	}
+	if err := WriteHeadHeaderHash(bc.chainDb, bc.currentHeader.Hash()); err != nil {
+		glog.Fatalf("failed to reset head header hash: %v", err)
+	}
+}
+
+// Rollback is designed to remove a chain of links from the database that aren't
+// certain enough to be valid.
+func (self *HeaderChain) Rollback(chain []common.Hash) {
+	for i := len(chain) - 1; i >= 0; i-- {
+		hash := chain[i]
+
+		if self.currentHeader.Hash() == hash {
+			self.currentHeader = self.GetHeader(self.currentHeader.ParentHash)
+			WriteHeadHeaderHash(self.chainDb, self.currentHeader.Hash())
+		}
+	}
+}
+
+// SetGenesis sets a new genesis block header for the chain
+func (self *HeaderChain) SetGenesis(head *types.Header) {
+	self.genesisHeader = head
+}
+
+// HdrValidator is responsible for validating block headers
+//
+// HdrValidator implements HeaderValidator.
+type HdrValidator struct {
+	bc  *HeaderChain // Canonical header chain
+	Pow pow.PoW      // Proof of work used for validating
+}
+
+// NewBlockValidator returns a new block validator which is safe for re-use
+func NewHdrValidator(chain *HeaderChain, pow pow.PoW) *HdrValidator {
```

Comment:
Please update the comment here. Also could we figure out a more decent name for this structr? `Hdr` doesn'tcut it in a modern codebase :P